ta/pkcs11/src/pkcs11_token.c

```diff
@@ -701,3 +701,621 @@ uint32_t entry_ck_session_info(struct pkcs11_client *client,
+
+		update_persistent_db(token);
+
+		if (token->db_main->flags & PKCS11_CKFT_USER_PIN_LOCKED)
+			return PKCS11_CKR_PIN_LOCKED;
+
+		return PKCS11_CKR_PIN_INCORRECT;
+	}
+
+	if (token->db_main->user_pin_count) {
+		token->db_main->user_pin_count = 0;
+
+		update_persistent_db(token);
+	}
+
+	if (token->db_main->flags & (PKCS11_CKFT_USER_PIN_COUNT_LOW |
+				     PKCS11_CKFT_USER_PIN_FINAL_TRY)) {
+		token->db_main->flags &= ~(PKCS11_CKFT_USER_PIN_COUNT_LOW |
+					   PKCS11_CKFT_USER_PIN_FINAL_TRY);
+
+		update_persistent_db(token);
+	}
+
+	return PKCS11_CKR_OK;
+}
+
+uint32_t entry_ck_set_pin(struct pkcs11_client *client,
+			  uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	struct pkcs11_session *session = NULL;
+	enum pkcs11_rc rc = PKCS11_CKR_OK;
+	struct serialargs ctrlargs = { };
+	uint32_t session_handle = 0;
+	uint32_t old_pin_size = 0;
+	TEE_Param *ctrl = params;
+	uint32_t pin_size = 0;
+	void *old_pin = NULL;
+	void *pin = NULL;
+
+	if (!client || ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get(&ctrlargs, &session_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &old_pin_size, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &pin_size, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_get_ptr(&ctrlargs, &old_pin, old_pin_size);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get_ptr(&ctrlargs, &pin, pin_size);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	session = pkcs11_handle2session(session_handle, client);
+	if (!session)
+		return PKCS11_CKR_SESSION_HANDLE_INVALID;
+
+	if (!pkcs11_session_is_read_write(session))
+		return PKCS11_CKR_SESSION_READ_ONLY;
+
+	if (pkcs11_session_is_so(session)) {
+		if (!(session->token->db_main->flags &
+		      PKCS11_CKFT_TOKEN_INITIALIZED))
+			return PKCS11_CKR_GENERAL_ERROR;
+
+		rc = check_so_pin(session, old_pin, old_pin_size);
+		if (rc)
+			return rc;
+
+		IMSG("PKCS11 session %"PRIu32": set PIN", session_handle);
+
+		return set_pin(session, pin, pin_size, PKCS11_CKU_SO);
+	}
+
+	if (!(session->token->db_main->flags &
+	      PKCS11_CKFT_USER_PIN_INITIALIZED))
+		return PKCS11_CKR_GENERAL_ERROR;
+
+	rc = check_user_pin(session, old_pin, old_pin_size);
+	if (rc)
+		return rc;
+
+	IMSG("PKCS11 session %"PRIu32": set PIN", session_handle);
+
+	return set_pin(session, pin, pin_size, PKCS11_CKU_USER);
+}
+
+static void session_login_user(struct pkcs11_session *session)
+{
+	struct pkcs11_client *client = session->client;
+	struct pkcs11_session *sess = NULL;
+
+	TAILQ_FOREACH(sess, &client->session_list, link) {
+		if (sess->token != session->token)
+			continue;
```

Comment:
This does assume that we eventually find the right one. Otherwise simply the last session in the list is used.
Maybe we should add
```
if (sess->token != session->token) return;
```
here to avoid manipulating the state of a wrong session (and possibly elevate privileges)

Reply:
My bad. Got the scope of the FOREACH wrong while reading.